server/src/main/java/org/activityinfo/server/endpoint/akvo/flow/AkvoFlowFeed.java

```diff
@@ -0,0 +1,125 @@
+import static org.activityinfo.model.resource.Resources.toJson;
+
+public class AkvoFlowFeed implements FeedService {
+    final private ResourceId parameterFormClassId;
+    final private ResourceLocatorSync locator;
+
+    public AkvoFlowFeed(ResourceId parameterFormClassId, ResourceLocatorSync locator) {
+        this.parameterFormClassId = parameterFormClassId;
+        this.locator = locator;
+    }
+
+    @Override
+    public FormClass getParameterFormClass() {
+        return locator.getFormClass(parameterFormClassId);
+    }
+
+    @Override
+    public void updateFeed(FormClass formClass, FormInstance parameters) {
+        try (AkvoFlow akvoFlow = initialize(parameters)) {
+            updateFeed(akvoFlow, formClass, parameters);
+        }
+    }
+
+    private void updateFeed(AkvoFlow akvoFlow, FormClass formClass, FormInstance parameters) {
+        List<FormInstance> formInstances = Lists.newArrayList();
+        ResourceId formClassId = formClass.getId(), timestampId = null;
+        Map<String, FormField> formFields = Maps.newHashMap();
+
+        for (FormField formField : getParameterFormClass().getFields()) {
+            if ("timestamp".equals(formField.getCode()) && parameters.getString(formField.getId()) != null) {
+                timestampId = formField.getId();
+            }
+        }
+
+        if (timestampId == null) {
+            throw new IllegalStateException("ParameterFormClass is invalid");
+        }
+
+        for (FormField formField : formClass.getFields()) {
+            formFields.put(formField.getCode(), formField);
+        }
+
+        for (SurveyInstance instance : akvoFlow.getSurveyInstances(parameters, timestampId)) {
+            FormInstance formInstance = new FormInstance(CuidAdapter.newLegacyFormInstanceId(formClassId), formClassId);
+            Date start = new Date(instance.collectionDate);
+            Date end = new Date(instance.collectionDate + instance.surveyalTime * 1000L);
+
+            formInstance.set(CuidAdapter.field(formClassId, CuidAdapter.START_DATE_FIELD), start);
+            formInstance.set(CuidAdapter.field(formClassId, CuidAdapter.END_DATE_FIELD), end);
+
+            for (QuestionAnswer questionAnswer : akvoFlow.getQuestionAnswers(instance.keyId)) {
+                FormField formField = formFields.get(questionAnswer.textualQuestionId);
+
+                if (formField == null) throw new IllegalStateException("Can't get " + questionAnswer.textualQuestionId);
+
+                for (EnumItem enumItem : ((EnumType) formField.getType()).getValues()) {
+                    if (enumItem.getLabel().equals(questionAnswer.value)) {
+                        formInstance.set(formField.getId(), new EnumValue(enumItem.getId()));
+                    }
+                }
+            }
+
+            formInstances.add(formInstance);
+        }
+
+        for (FormInstance formInstance : formInstances) {
+            getDefaultQueue().add(withUrl("/tasks/persist")
```

Comment:
Actually, it looks like you're independently queuing each form instance? This quite problematic as it offers no gaurantees on consistency: it could very well happen that instances submitted later to Flow arrive earlier in activityinfo. And if there are fatal errors, then you could end up with holes in the synchronization. 

What I think is a better idea is to do all the persists within a single worker task. If there is a timeout or transient problem persisting the data, then stop, save the time of the last successfully persisted item and reschedule the task for later (or it will simply run again on the cron schedule)

This will ensure that AI remains consistent with Flow.